lib/streamlit/hello.py

```diff
@@ -17,55 +17,172 @@
+    array[:, :, 1].fill(green)
+    array[:, :, 2].fill(blue)
+    st.markdown("#### RGB color (%d, %d, %d)" % (red, green, blue))
+    st.image(array)
 
-        # Update status text.
-        status_text.text("The latest random number is: %s" % new_rows[-1, 0])
 
-        # Append data to the chart.
-        chart.add_rows(new_rows)
+# Data for demo_5 derived from
+# https://www.kaggle.com/rounakbanik/the-movies-dataset
+DATASET_URL = (
+    "https://streamlit-demo-data.s3-us-west-2.amazonaws.com/movies-revenue.csv.gz"
+)
 
-        # Pretend we're doing some computation that takes time.
-        time.sleep(0.1)
 
-    status_text.text("Done!")
-    st.balloons()
+def demo_5():
+    """
+    Discover the gross revenue of a movie of your liking!
+    Note that the network loading and preprocessing of the data is cached
+    by using Streamlit st.cache annotation.
+    """
+    import pandas as pd
+
+    @st.cache
+    def load_dataframe_from_url():
+        return pd.read_csv(DATASET_URL).transform(
+            {"title": lambda x: x, "revenue": lambda x: round(x / 1000 / 1000)}
+        )
+
+    df = load_dataframe_from_url()
+    movie = st.text_input("Search movie titles")
+    df = df[df.title.str.contains(movie, case=False)].sort_values(
+        ascending=False, by="revenue"
+    )
+    st.markdown("#### Result dataframe for %d movie(s)" % df.shape[0])
+    st.dataframe(df.rename(columns={"title": "Title", "revenue": "Revenue [$1M]"}))
+
 
+DEMOS = OrderedDict(
+    {
+        "---": intro,
+        "Basic Text": demo_1,
+        "Simple Interaction": demo_2,
+        "Animation": demo_3,
+        "Sidebar": demo_4,
+        "Caching": demo_5,
+    }
+)
+
+
+def run():
+    st.title("Welcome to Streamlit!")
     st.write(
         """
-        PS: Want to know how we did this?
-        [You can learn about it here.](https://streamlit.io/secret/docs/core_mechanics.html)
-    """
+        Streamlit is the best way to create bespoke apps.
+        """
     )
 
+    demo_name = st.selectbox("Choose a demo", list(DEMOS.keys()), 0)
+    demo = DEMOS[demo_name]
+
+    if demo_name != "---":
+        st.markdown("## %s Demo" % demo_name)
+        st.write(inspect.getdoc(demo))
+        st.write("### Code")
+        sourcelines, n_lines = inspect.getsourcelines(demo)
+        sourcelines = reset_indentation(remove_docstring(sourcelines))
+        st.code("".join(sourcelines))
+        if st.checkbox("Run %s Demo" % demo_name):
+            st.write("---\n### Execution")
+            demo()
+    else:
+        demo()
+
+
+# This function parses the lines of the function and removes the docstring
+# if found.
+def remove_docstring(lines):
+    if len(lines) < 3 and '"""' not in lines[1]:
+        return lines
+    #  lines[2] is the first line of the docstring, past the inital """
+    index = 2
+    while '"""' not in lines[index]:
+        index += 1
+        # limit to ~100 lines
+        if index > 100:
+            return lines
```

Comment:
Isn't this also including the docstring in `lines` when it returns?

Reply:
Yes, I am bailing out if the docstring is too big or I haven't found a closing `"""`.